### ISSUE TITLE
Watch does not call callback on cold cache start

### ISSUE THREAD

**Issue by vishal-p** [apollographql/apollo-ios#99]
Hi
I'm using Apollo-iOS beta0.6.0.
Is the following behaviour expected?

1. No cached data. App is started.
2. Initialise a callback using watch with  `returnCacheDataDontFetch`

```
apollo.watch(query: spacesQuery, cachePolicy: .returnCacheDataDontFetch , resultHandler: { (result , error ) in
            if let error = error {
                print(#function, "ERROR | An error occured: \(error)")
                return
            }
            
            guard let edges  = result?.data?.node?.asAccount?.spaces?.edges else {                  
                print(#function, "ERROR | Could not retrieve trainer")
                return
            }
           
            completionBlock(result,nil)
        })
```

3. Query the graphql endpoint using fetch `fetchIgnoringCacheData`
```

     apollo.fetch(query: spacesQuery, cachePolicy: .fetchIgnoringCacheData, resultHandler: { (result , 
         error ) in
            if let error = error {
                print(#function, "ERROR | An error occured: \(error)")
                return
            }
            guard let _ = result?.data?.node?.asAccount?.spaces?.edges else {
                print(#function, "ERROR | Could not retrieve trainer")
                return
            }
        })
```

4. Watch callback was not fired
5. On next app launch watch is called correctly once we've data in cache.
If the above is correct, what would be the recommended solution so that watch fires correctly after a fetch call is made when data is initially retrieved.

**Comment by martijnwalraven** [comment]
Do you store the `GraphQLQueryWatcher` you get from `apollo.watch()` somewhere? If not, the watcher will be immediately deinitialized.

See [FrontPage app](https://github.com/apollographql/frontpage-ios-app/blob/master/FrontPage/PostListViewController.swift#L28-L39) for an example.

**Comment by vishal-p** [comment]
Yes I'm storing `GraphQLQueryWatcher` in class level variable. It calls once I've fetched data from network.

**Comment by martijnwalraven** [comment]
Ah, I think this is actually expected behavior (although we may want to rethink it). Basically, a watcher works by watching for a set of dependent keys. But it can only get these keys if it gets at least one valid response (either from the cache or from the network). Because your `watch` uses `.returnCacheDataDontFetch`, and the cache is initially empty, it will never get a valid response in that case.

**Comment by vishal-p** [comment]
Thanks for replying. It would be good if watcher notifies on any cache updates. As I see above is valid case, I'm using watcher only for loading/watching cached data changes, and fetch gets data from network.

As an alternative approach I'm thinking something like this. 
1. `fetch` using `returnCacheDataDontFetch` and check for any cached data.
2. if there is no data then fetch from network  and on successful result render UI and setup watcher.
3. Next time watcher will be automatically called and updates UI.
